File: flutter_app/lib/services/membership_service.dart
import '../models/membership.dart';
import 'api_service.dart';

/// Membership service for managing memberships
class MembershipService {
  final ApiService _api = ApiService();

  /// Get all memberships
  Future<List<Membership>> getAllMemberships() async {
    final response = await _api.get('/api/memberships');
    return (response as List).map((json) => Membership.fromJson(json)).toList();
  }

  /// Get membership by ID
  Future<Membership> getMembershipById(int id) async {
    final response = await _api.get('/api/memberships/$id');
    return Membership.fromJson(response);
  }

  /// Get active memberships
  Future<List<Membership>> getActiveMemberships() async {
    final response = await _api.get('/api/memberships/active');
    return (response as List).map((json) => Membership.fromJson(json)).toList();
  }

  /// Create membership (admin only)
  Future<Membership> createMembership(Membership membership) async {
    final response = await _api.post('/api/memberships', body: membership.toJson());
    return Membership.fromJson(response);
  }

  /// Update membership (admin only)
  Future<Membership> updateMembership(int id, Membership membership) async {
    final response = await _api.put('/api/memberships/$id', body: membership.toJson());
    return Membership.fromJson(response);
  }

  /// Delete membership (admin only)
  Future<void> deleteMembership(int id) async {
    await _api.delete('/api/memberships/$id');
  }
}

File: nodejs-backend/src/services/membership.service.js
/**
 * Membership Service
 * Handles membership-related business logic
 */

const { Membership } = require('../models');

class MembershipService {
  /**
   * Get all memberships
   */
  async getAllMemberships() {
    return await Membership.findAll({
      order: [['price', 'ASC']]
    });
  }

  /**
   * Get membership by ID
   */
  async getMembershipById(id) {
    return await Membership.findByPk(id);
  }

  /**
   * Get active memberships
   */
  async getActiveMemberships() {
    return await Membership.findAll({
      where: { is_active: true },
      order: [['price', 'ASC']]
    });
  }

  /**
   * Create new membership
   */
  async createMembership(membershipData) {
    return await Membership.create({
      name: membershipData.name,
      description: membershipData.description,
      price: membershipData.price,
      duration_days: membershipData.durationDays || 30,
      benefits: membershipData.benefits,
      is_active: membershipData.isActive !== undefined ? membershipData.isActive : true
    });
  }

  /**
   * Update membership
   */
  async updateMembership(id, membershipData) {
    const membership = await Membership.findByPk(id);
    if (!membership) {
      throw new Error('Membership not found');
    }

    if (membershipData.name !== undefined) membership.name = membershipData.name;
    if (membershipData.description !== undefined) membership.description = membershipData.description;
    if (membershipData.price !== undefined) membership.price = membershipData.price;
    if (membershipData.durationDays !== undefined) membership.duration_days = membershipData.durationDays;
    if (membershipData.benefits !== undefined) membership.benefits = membershipData.benefits;
    if (membershipData.isActive !== undefined) membership.is_active = membershipData.isActive;

    await membership.save();
    return membership;
  }

  /**
   * Delete membership
   */
  async deleteMembership(id) {
    const membership = await Membership.findByPk(id);
    if (!membership) {
      throw new Error('Membership not found');
    }

    await membership.destroy();
    return true;
  }
}

module.exports = new MembershipService();
